### PR TITLE
Define missing attribute methods from `method_missing`

### PR DESCRIPTION
**Pull request by byroot**
Ref: https://github.com/mastodon/mastodon/issues/27622

If applications don't eager load and use the old Rails 6.1 Marshal format, they can load Active Record instances from caches without calling `init_internals` hence attribute methods and aliases are nnot defined yet, leading to `NoMethodError`

Initially I was hopping to fully get rid of the `define_attribute_methods` call in `init_internals` in favor of this new method missing, as it would remove work from the happy path, unfortunately that isn't possible because if a generated method overrides a default method inherited from object, `method_missing` won't be called. e.g. `Kernel#format`

We may want to get rid of this extra code once we remove support for the 6.1 marshal format.

cc @ClearlyClaire
FYI: @nvasilevski

### CODE REVIEW

File: activerecord/lib/active_record/model_schema.rb [hunk 1]
```diff
@@ -524,9 +524,9 @@ def reset_column_information
       end
 
       def load_schema # :nodoc:
-        return if schema_loaded?
+        return false if schema_loaded?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/model_schema.rb:527]: Why is this returning false now?

Comment (byroot) [reply activerecord/lib/active_record/model_schema.rb:527]: To know if the schema was loaded or not. I didn't need it in the end, so I could revert this part.

**Review by rafaelfranca — APPROVED** [review]

File: activerecord/lib/active_record/model_schema.rb (hunk 1 above)
Comment (rafaelfranca) [reply activerecord/lib/active_record/model_schema.rb:527]: Is it to differentiate from `nil`? I can't see where `nil` could happen. If it is to match the `true` with `false` I think we should remove it, `nil` is false.

Comment (byroot) [reply activerecord/lib/active_record/model_schema.rb:527]: It was just to be more explicit. But yes I'll remove it.

**Comment by byroot** [comment]
Backported to 7-1-stable.

**Comment by skipkayhil** [comment]
@byroot did the backport not get pushed up? I'm not seeing it on `7-1-stable`

**Comment by byroot** [comment]
yeah I forgot to push 😫 . It's now at f7255ec27490580a41b21c71b54d2d2fec18c269

**Comment by rafaelfranca** [comment]
@byroot I think isn't backported yet. I don't see this code in the branch. https://github.com/rails/rails/commits/7-1-stable/activerecord/lib/active_record/model_schema.rb

There is a changelog entry for it, but no other code change went, and the commit message is wrong.

**Comment by byroot** [comment]
Hum, I must have borked my merge, I will have a look.

**Comment by stephankaag** [comment]
@byroot @rafaelfranca 
This change changed / broke the following scenario where AR tries to load the schema of a non-existing table. 

I noticed because [motor-admin-rails](https://github.com/motor-admin/motor-admin-rails) uses this pattern with the [auditable](https://github.com/collectiveidea/audited) gem.

Setup:

```ruby 
class Audit < ApplicationRecord
  belongs_to :auditable, polymorphic: true
end

class Article < ApplicationRecord
  has_many :audits, as: :auditable, class_name: "MotorAudit", inverse_of: :auditable
end

class MotorAudit < Audit
  self.table_name = 'motor_audits'
end

create_table :motor_audits do |t|
  t.column :auditable_id, :string
  t.column :auditable_type, :string
end

create_table :articles do |t|
  t.timestamps
end

```

Rails 7.1.3.2:
```
> article = Article.create
  TRANSACTION (0.1ms)  begin transaction
  Article Create (0.8ms)  INSERT INTO "articles" ("created_at", "updated_at") VALUES (?, ?) RETURNING "id"  [["created_at", "2024-02-28 09:09:01.693949"], ["updated_at", "2024-02-28 09:09:01.693949"]]
  TRANSACTION (0.1ms)  commit transaction
 => #<Article:0x0000000107e57588 id: 4, created_at: Wed, 28 Feb 2024 09:09:01.693949000 UTC +00:00, updated_at: Wed, 28 Feb 2024 09:09:01.693949000 UTC +00:00> 
> article.audits.create
  TRANSACTION (0.0ms)  begin transaction
  MotorAudit Create (0.2ms)  INSERT INTO "motor_audits" ("auditable_id", "auditable_type") VALUES (?, ?) RETURNING "id"  [["auditable_id", "4"], ["auditable_type", "Article"]]
  TRANSACTION (0.1ms)  commit transaction
 => #<MotorAudit:0x00000001078ffa40 id: 2, auditable_id: "4", auditable_type: "Article"> 
```

Rails 7.2.0.alpha:
```
> article = Article.create
  TRANSACTION (0.0ms)  begin transaction
  Article Create (0.8ms)  INSERT INTO "articles" ("created_at", "updated_at") VALUES (?, ?) RETURNING "id"  [["created_at", "2024-02-28 09:09:57.515359"], ["updated_at", "2024-02-28 09:09:57.515359"]]
  TRANSACTION (0.1ms)  commit transaction
 => #<Article:0x0000000105e944a8 id: 5, created_at: Wed, 28 Feb 2024 09:09:57.515359000 UTC +00:00, updated_at: Wed, 28 Feb 2024 09:09:57.515359000 UTC +00:00> 
> article.audits.create
ruby-3.2.2/bundler/gems/rails-649e99bf9267/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:484:in `table_structure': Could not find table 'audits' (ActiveRecord::StatementInvalid)
```

**Comment by byroot** [comment]
If the table doesn't exist, shouldn't `Audit` have `self.abstract_class = true`?

If that doesn't solve your problem, please open an issue with a repro script, and I'll figure a solution.

**Comment by stephankaag** [comment]
> If the table doesn't exist, shouldn't `Audit` have `self.abstract_class = true`?
> 
> If that doesn't solve your problem, please open an issue with a repro script, and I'll figure a solution.

Not sure. But at least I wanted to mention that the behaviour is now changed.

**Comment by byroot** [comment]
> But at least I wanted to mention that the behaviour is now changed.

Of course, but to be clear, if that behavior wasn't supported nor documented in the first place but just happened to work by accident, it's not a concern.

**Pull request merged**